### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by DanCh11**
Fine-tuned the algorithm and reviewed the evaluation metrics, then up loaded to mlflow

### CODE REVIEW

File: sc_wrangling/dictionary_transformation.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from pandas import DataFrame
+
+
+def transform_series_to_dictionary(data_frame: DataFrame, key_column: str, value_column: str):
```
Comment (costezki) [inline sc_wrangling/dictionary_transformation.py:4]: well written function:
short
clean
documented.

I would propose to rename it however into something like: `series_pair_to_dict`

""""
assuming that we want to transform two Dataframe columns into a dictionary, where the first column acts as unique keys and the second acts as values. Note that the key uniqueness is not verified. If there are multiple records with the same key, then only the latest (value) is considered .... 
""""

**Review by costezki — APPROVED** [review]

File: notebooks/pwdb/algorithms/KNN/law2vec/target_groups_classes.ipynb [hunk 2]
```diff
@@ -2,7 +2,7 @@
  "cells": [
```
Comment (costezki) [inline notebooks/pwdb/algorithms/KNN/law2vec/target_groups_classes.ipynb:2]: <p>Probably a good practice is to run the entire notebook before commit. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (costezki) [inline notebooks/pwdb/algorithms/KNN/law2vec/target_groups_classes.ipynb:2]: <p>we shall soon the other evaluation metrics (Area under ROC, MCC, etc..).</p><p> </p><p>The Mean Absolute Error and Mean Squared Error have the same value. I doubt they are the same. </p><p>Please check the <code>model_evaluation_metrics</code>  function. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (costezki) [inline notebooks/pwdb/algorithms/KNN/law2vec/target_groups_classes.ipynb:2]: <p>It is better to define CONSTANTS for parameter values so that you don't need to remember and  'copy' that value, but simply refer to that value whenever you need it. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

File: notebooks/pwdb/algorithms/KNN/law2vec/target_groups_subclasses.ipynb [hunk 3]
```diff
@@ -23,6 +23,11 @@
     "TABLE_PATH = FOLDER / 'data/pickle/df_columns_labels.pkl'\n",
```
Comment (costezki) [inline notebooks/pwdb/algorithms/KNN/law2vec/target_groups_subclasses.ipynb:23]: <p>I would recommend the <code>n_neighbors=3</code>, <code>q=2</code>   to be defined as CONSTANTS in the beginning of the notebook, and then refer to them both in model training and in mlflow logging. Later these constants will become Papermill parameters. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (costezki) [inline notebooks/pwdb/algorithms/KNN/law2vec/target_groups_subclasses.ipynb:23]: <p>this is still copied manually. You have the function that transforms DF into  dict. </p><p>Alternatively you may simply have the <code>model_evaluation_metrics</code>  function return the <code>dict</code>  instead of DataFrame. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

File: notebooks/pwdb/algorithms/KNN/target_groups_classes.ipynb [hunk 4]
```diff
@@ -2,7 +2,7 @@
  "cells": [
```
Comment (costezki) [inline notebooks/pwdb/algorithms/KNN/target_groups_classes.ipynb:2]: <p>This class, I believe is for calculating the Document embedding. </p><p>It should NOT be in the notebook but in an external python module named for example <code>embedding_handlers.py</code>. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (costezki) [inline notebooks/pwdb/algorithms/KNN/target_groups_classes.ipynb:2]: <p>(hyper-)parameters shall be constants defined at the beginning of the notebook. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (costezki) [inline notebooks/pwdb/algorithms/KNN/target_groups_classes.ipynb:2]: <p>how is this score calculated? What is its name?</p><p>Can we have it included into the <code>model_evaluation_metrics</code>  ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (costezki) [inline notebooks/pwdb/algorithms/KNN/target_groups_classes.ipynb:2]: <p>parameters shall be all, defined as constants in the beginning of the notebook and reused whenever needed. </p><p>We shall try avoid as much as possible using numbers directly in the code. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

File: notebooks/pwdb/algorithms/KNN/target_groups_subclasses.ipynb [hunk 5]
```diff
@@ -2,7 +2,7 @@
  "cells": [
```
Comment (costezki) [inline notebooks/pwdb/algorithms/KNN/target_groups_subclasses.ipynb:2]: <p>constants ...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (costezki) [inline notebooks/pwdb/algorithms/KNN/target_groups_subclasses.ipynb:2]: <p>constants ....</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

File: notebooks/pwdb/algorithms/SVM/category.ipynb [hunk 6]
```diff
@@ -2,7 +2,7 @@
  "cells": [
```
Comment (costezki) [inline notebooks/pwdb/algorithms/SVM/category.ipynb:2]: <p>constants ...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

File: notebooks/pwdb/algorithms/SVM/law2vec/subcategory.ipynb [hunk 7]
```diff
@@ -2,7 +2,7 @@
  "cells": [
```
Comment (costezki) [inline notebooks/pwdb/algorithms/SVM/law2vec/subcategory.ipynb:2]: <p>Once a code is written it shall be reused. We should NEVER rewrite a class or a function. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

File: notebooks/pwdb/algorithms/SVM/law2vec/type_of_measure.ipynb [hunk 8]
```diff
@@ -2,7 +2,7 @@
  "cells": [
```
Comment (costezki) [inline notebooks/pwdb/algorithms/SVM/law2vec/type_of_measure.ipynb:2]: <p>Do we have document embeddings in the KNN notebooks? </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/meaningfy-ws/ml-experiments/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

File: notebooks/pwdb/algorithms/KNN/law2vec/target_groups_classes.ipynb (hunk 2 above)
Comment (DanCh11) [reply notebooks/pwdb/algorithms/KNN/law2vec/target_groups_classes.ipynb:2]: <p>The function seems to be correct. In other experiments, Mean Absolute Error and Mean Squared Error are different</p>

File: notebooks/pwdb/algorithms/KNN/target_groups_classes.ipynb (hunk 4 above)
Comment (DanCh11) [reply notebooks/pwdb/algorithms/KNN/target_groups_classes.ipynb:2]: <p>the score() method calculates the difference between X_train and y_train (an accuracy measure).</p>

Comment (DanCh11) [reply notebooks/pwdb/algorithms/KNN/target_groups_classes.ipynb:2]: <p>I need to inspect how we can include it into our <code>model_evaluation_metrics</code>, because it's the algorithm's method</p>

Comment (costezki) [reply notebooks/pwdb/algorithms/KNN/target_groups_classes.ipynb:2]: `score(X, y, sample_weight=None)[source]`

Return the mean accuracy on the given test data and labels.

In multi-label classification, this is the subset accuracy which is a harsh metric since you require for each sample that each label set be correctly predicted.

[from sklearn]
